internal/crd/demand_informer.go

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2019 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crd
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	demandapi "github.com/palantir/k8s-spark-scheduler-lib/pkg/apis/scaler/v1alpha1"
+	ssinformers "github.com/palantir/k8s-spark-scheduler-lib/pkg/client/informers/externalversions"
+	"github.com/palantir/k8s-spark-scheduler-lib/pkg/client/informers/externalversions/scaler/v1alpha1"
+	"github.com/palantir/pkg/retry"
+	werror "github.com/palantir/witchcraft-go-error"
+	"github.com/palantir/witchcraft-go-logging/wlog/svclog/svc1log"
+	"github.com/palantir/witchcraft-go-logging/wlog/wapp"
+	apiextensionsclientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	clientcache "k8s.io/client-go/tools/cache"
+)
+
+const (
+	informerSyncRetryCount          = 5
+	informerSyncTimeout             = 2 * time.Second
+	informerSyncRetryInitialBackoff = 500 * time.Millisecond
+)
+
+// LazyDemandInformer checks for Demand CRD existence and creates a
+// demand informer if it exists.
+type LazyDemandInformer struct {
+	informerFactory     ssinformers.SharedInformerFactory
+	apiExtensionsClient apiextensionsclientset.Interface
+	ready               chan struct{}
+	informer            v1alpha1.DemandInformer
+	lock                sync.RWMutex
+}
+
+// NewLazyDemandInformer constructs a new LazyDemandInformer instance
+func NewLazyDemandInformer(
+	informerFactory ssinformers.SharedInformerFactory,
+	apiExtensionsClient apiextensionsclientset.Interface) *LazyDemandInformer {
+	return &LazyDemandInformer{
+		informerFactory:     informerFactory,
+		apiExtensionsClient: apiExtensionsClient,
+		ready:               make(chan struct{}),
+	}
+}
+
+// Informer returns the informer instance if it is initialized, returns false otherwise
+func (ldi *LazyDemandInformer) Informer() (v1alpha1.DemandInformer, bool) {
+	ldi.lock.RLock()
+	defer ldi.lock.RUnlock()
+	select {
+	case <-ldi.Ready():
+		return ldi.informer, true
+	default:
+		return nil, false
+	}
+}
+
+// Ready returns a channel that will be closed when the informer is initialized
+func (ldi *LazyDemandInformer) Ready() <-chan struct{} {
+	return ldi.ready
+}
+
+// Run starts the goroutine to check for the existence of the demand CRD,
```

Comment:
`starts the goroutine to check for the existence of the demand CRD`, but seems like it's happenning synchronously?

Reply:
it does the first retry synchronously, if failed starts the goroutine to check periodically